jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/StateBasedNodeInstance.java

```diff
@@ -345,9 +347,20 @@ private void triggerTimer(TimerInstance timerInstance) {
         for (Map.Entry<Timer, DroolsAction> entry : getEventBasedNode().getTimers().entrySet()) {
             if (entry.getKey().getId() == timerInstance.getTimerId()) {
                 executeAction((Action) entry.getValue().getMetaData("Action"));
+
+                // self remove timer instance from this node as it will remove itself 
+                // from timer service once is triggered and there is no next
+                if (!(timerInstance.getJobHandle() instanceof DefaultJobHandle)) {
+                    return;
+                }
+                DefaultTimerJobInstance defaultTimerInstance = (DefaultTimerJobInstance) ((DefaultJobHandle) timerInstance.getJobHandle()).getTimerJobInstance();
```

Comment:
Is the casting to DefaultTimerJobInstance needed? For example, replacing _defaultTimerInstance_ in the next line by 
_((DefaultJobHandle) timerInstance.getJobHandle()).getTimerJobInstance()_

Reply:
yes it is needed JobHandle does not have method getTimerJobInstance

Reply:
I mean `TimerJobInstance` interface already has `getTrigger` method, so there's no need to cast to `DefaultTimerJobInstance`, if you don't create an intermediate object (but it's not important, just nitpicking)

```suggestion
if (((DefaultJobHandle) timerInstance.getJobHandle()).getTimerJobInstance().getTrigger().hasNextFireTime() == null) {
```